File: 02. Algorithms/09. Dynamic Programming/008. Nth Tribonacci Number/code.cpp
class Solution {
public:
    int tribonacci(int n) {
        if(n==0 || n==1)
            return n;
        
        if(n==2)
            return 1;
        
        int dp[n+1];
        
        dp[0]=0;dp[1]=1;dp[2]=1;
        
        for(int i=3;i<=n;i++)
            dp[i] = dp[i-1]+dp[i-2]+dp[i-3] ;
        
        return dp[n];
    }
};

File: 02. Algorithms/09. Dynamic Programming/008. Nth Tribonacci Number/code.js
/**
 * Recursion | DP
 * https://leetcode.com/problems/n-th-tribonacci-number
 * 
 * Time O(n) | Space O(n)
 * @param {number} n
 * @return {number}
 */
var tribonacci = function(n) {

    const cache = new Map();
    if(n === 1 || n === 2) return 1;
    if(n === 0) return 0;
    const dfs = (n) => {
      
      if(cache.has(n)) return cache.get(n);
      if(n === 2 || n === 1) return 1;
      if(n === 0) return 0;
      
      const result = dfs(n-3) + dfs(n-2) + dfs(n-1);
      cache.set(n, result);
      return result;     
    }

    return dfs(n);
};
